src/java/org/apache/cassandra/index/sai/disk/v3/BinaryTree.java

```diff
@@ -0,0 +1,358 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.disk.v3;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.agrona.collections.IntArrayList;
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.lucene.index.PointValues;
+import org.apache.lucene.store.ByteBuffersDataOutput;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.util.ArrayUtil;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.BytesRefArray;
+import org.apache.lucene.util.BytesRefBuilder;
+import org.apache.lucene.util.MathUtil;
+
+/**
+ * Binary tree based on the Lucene 8.x kdtree.
+ * <p>
+ * The index is not prefix compressed yet.
+ */
+public class BinaryTree
+{
+    public interface IntersectVisitor
+    {
+        /** Called for non-leaf cells to test how the cell relates to the query, to
+         *  determine how to further recurse down the tree. */
+        PointValues.Relation compare(BytesRef minPackedValue, BytesRef maxPackedValue);
```

Comment:
I notice that this interface already exists in both `BKDTree.java` and `MutatableOneDimPointValues.java`. Can we put it in one location?